@app/server/src/plugins/PgTypeEmailPlugin.ts

```diff
@@ -0,0 +1,149 @@
+      );
+
+      if (!emailType) {
+        return build;
+      }
+
+      const emailTypeName = build.inflection.builtin("Email");
+
+      const GraphQLEmailType = makeGraphQLEmailType(
+        build as Build,
+        emailTypeName
+      );
+
+      // Now register the Email type with the type system for both output and input.
+      pgRegisterGqlTypeByTypeId(emailType.id, () => GraphQLEmailType);
+      pgRegisterGqlInputTypeByTypeId(emailType.id, () => GraphQLEmailType);
+
+      // Finally we must tell the system how to translate the data between PG-land and JS-land:
+      pg2GqlMapper[emailType.id] = {
+        // Turn string (from node-postgres) into email: no-op
+        map: (email: string) => email,
+        // When unmapping we need to convert back to SQL framgent
+        unmap: (email: string) =>
+          sql.fragment`(${sql.value(email)}::${sql.identifier(
+            emailType.namespaceName,
+            emailType.name
+          )})`,
+      };
+
+      return build;
+    },
+    ["PgTypeEmail"],
+    [],
+    ["PgTypes"]
+  );
+
+  /* End of email type */
+} as Plugin);
+
+function makeGraphQLEmailType(build: Build, emailTypeName: string) {
+  const {
+    graphql: { GraphQLScalarType, Kind },
+  } = build;
+  function parseValue(obj: unknown): string {
+    if (!(typeof obj === "string")) {
+      throw new TypeError(
+        `This is not a valid ${emailTypeName} object, it must be a string.`
+      );
+    }
+    if (!isValidEmail(obj)) {
+      throw new TypeError(
+        `This is not a properly formatted ${emailTypeName} object.`
+      );
+    }
+    return obj;
+  }
+
+  const parseLiteral: import("graphql").GraphQLScalarLiteralParser<any> = (
+    ast,
+    variables
+  ) => {
+    switch (ast.kind) {
+      case Kind.STRING: {
+        const email = ast.value;
+        if (!isValidEmail(email)) {
+          throw new TypeError(
+            `This is not a properly formatted ${emailTypeName} object.`
+          );
+        }
+        return email;
+      }
+
+      case Kind.NULL:
+        return null;
+
+      case Kind.VARIABLE: {
+        const name = ast.name.value;
+        const email = variables ? variables[name] : undefined;
+        if (!isValidEmail(email)) {
+          throw new TypeError(
+            `This is not a properly formatted ${emailTypeName} object.`
+          );
+        }
+        return email;
+      }
+
+      default:
+        return undefined;
+    }
+  };
+
+  const scope: ScopeGraphQLScalarType = { isEmailScalar: true };
+  const GraphQLEmailType = build.newWithHooks(
+    GraphQLScalarType,
+    {
+      name: emailTypeName,
+      description:
+        "An address in the electronic mail system. Email addresses such as `John.Smith@example.com` are made up of a local-part, followed by an `@` symbol, followed by a domain.",
+      serialize: (email: string) => email,
+      parseValue,
+      parseLiteral,
+    },
+    scope
+  );
+
+  return GraphQLEmailType;
+}
```

Comment:
@singingwolfboy  It seems to me that this plugin adds the `Email` type, with a comment, and performs validation on it. However we already perform validation on the type using the domain constraint you added; so I'm not sure if these 149 lines of code add much additional value? If this plugin were removed, would the following be sufficient?

```sql
create domain app_public.email as citext check(VALUE ~ '[^@]+@[^@]+\.[^@]+');
comment on domain app_public.email is
  'An address in the electronic mail system. Email addresses such as `John.Smith@example.com` are made up of a local-part, followed by an `@` symbol, followed by a domain.';
```

Reply:
I’m cool with the custom scalar and totally understand the reasoning; for some reason I was expecting creating a custom domain to result in a custom scalar being created. Seems maybe I never implemented that https://github.com/graphile/graphile-engine/blob/v4/packages/graphile-build-pg/src/plugins/PgTypesPlugin.js#L844